command/debug/debug.go

```diff
@@ -0,0 +1,613 @@
+	debugArchiveExtension = ".tar.gz"
+)
+
+func New(ui cli.Ui, shutdownCh <-chan struct{}) *cmd {
+	ui = &cli.PrefixedUi{
+		OutputPrefix: "==> ",
+		InfoPrefix:   "    ",
+		ErrorPrefix:  "==> ",
+		Ui:           ui,
+	}
+
+	c := &cmd{UI: ui, shutdownCh: shutdownCh}
+	c.init()
+	return c
+}
+
+type cmd struct {
+	UI    cli.Ui
+	flags *flag.FlagSet
+	http  *flags.HTTPFlags
+	help  string
+
+	shutdownCh <-chan struct{}
+
+	// flags
+	interval time.Duration
+	duration time.Duration
+	output   string
+	archive  bool
+	capture  []string
+	client   *api.Client
+	// validateTiming can be used to skip validation of interval, duration. This
+	// is primarily useful for testing
+	validateTiming bool
+}
+
+func (c *cmd) init() {
+	c.flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+	defaultFilename := fmt.Sprintf("consul-debug-%d", time.Now().Unix())
+
+	c.flags.Var((*flags.AppendSliceValue)(&c.capture), "capture",
+		"One or more types of information to capture. This can be used "+
+			"to capture a subset of information, and defaults to capturing "+
+			"everything available. Possible information for capture: "+
+			"This can be repeated multiple times.")
+	c.flags.DurationVar(&c.interval, "interval", debugInterval,
+		fmt.Sprintf("The interval in which to capture dynamic information such as "+
+			"telemetry, and profiling. Defaults to %s.", debugInterval))
+	c.flags.DurationVar(&c.duration, "duration", debugDuration,
+		fmt.Sprintf("The total time to record information. "+
+			"Defaults to %s.", debugDuration))
+	c.flags.BoolVar(&c.archive, "archive", true, "Boolean value for if the files "+
+		"should be archived and compressed. Setting this to false will skip the "+
+		"archive step and leave the directory of information on the relative path.")
+	c.flags.StringVar(&c.output, "output", defaultFilename, "The path "+
+		"to the compressed archive that will be created with the "+
+		"information after collection.")
+
+	c.http = &flags.HTTPFlags{}
+	flags.Merge(c.flags, c.http.ClientFlags())
+	// TODO do we need server flags?
+	flags.Merge(c.flags, c.http.ServerFlags())
+	c.help = flags.Usage(help, c.flags)
+
+	c.validateTiming = true
+}
+
+func (c *cmd) Run(args []string) int {
+	if err := c.flags.Parse(args); err != nil {
+		c.UI.Error(fmt.Sprintf("Error parsing flags: %s", err))
+		return 1
+	}
+
+	// Connect to the agent
+	client, err := c.http.APIClient()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error connecting to Consul agent: %s", err))
+		return 1
+	}
+	c.client = client
+
+	version, err := c.prepare()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Capture validation failed: %v", err))
+		return 1
+	}
+
+	archiveName := c.output
+	// Show the user the final file path if archiving
+	if c.archive {
+		archiveName = archiveName + debugArchiveExtension
+	}
+
+	c.UI.Output("Starting debugger and capturing static information...")
+
+	// Output metadata about target agent
+	c.UI.Info(fmt.Sprintf(" Agent Address: '%s'", "TODO"))
```

Comment:
What you probably want here is to use the (resolved) `-advertise` address for the agent since that is the one that will uniquely identify it on the network in gossip. We should return that in the HOST endpoint if we don't already - it's the same thing we configured Serf with so presumably it's in config somewhere and easy to return.

Reply:
Hm yeah I could do that. I was trying to use the one the API client was configured with in this context so I could show the user, but using the resolved one I guess does make sense. Could they be connecting to it via some different address, though?

Reply:
> I was trying to use the one the API client was configured with

That was my first thought but then in a lot of cases it's just going to be `127.0.0.1` which is not really useful? We don't really recommend exposing the HTTP API outside of localhost even on servers so for a best-practice deployment the operator presumably has to run the debug command on localhost anyway so the agent IP they are using is not very interesting.

I'm assuming the value in showing this is that it lets us correlate with logs which will show the external/advertise address in general.

Reply:
Yeah, I think I will just remove this. If the agent is unreachable, that address will be in the error output following the initial request, and if it is reachable we have all we need in the archive.